safenode/src/client/file_apis.rs

```diff
@@ -0,0 +1,304 @@
+            return self.seek(data_map, position, length).await;
+        }
+
+        // The error above is ignored to avoid leaking the storage format detail of SmallFiles and LargeFiles.
+        // The basic idea is that we're trying to deserialize as one, and then the other.
+        // The cost of it is that some errors will not be seen without a refactor.
+        let mut bytes = chunk.value().clone();
+
+        let _ = bytes.split_to(position);
+        bytes.truncate(length);
+
+        Ok(bytes)
+    }
+
+    /// Directly writes [`Bytes`] to the network in the
+    /// form of immutable chunks, without any batching.
+    #[instrument(skip(self, bytes), level = "debug")]
+    pub async fn upload(&self, bytes: Bytes) -> Result<ChunkAddress> {
+        self.upload_bytes(bytes, false).await
+    }
+
+    /// Directly writes [`Bytes`] to the network in the
+    /// form of immutable chunks, without any batching.
+    /// It also attempts to verify that all the data was uploaded to the network before returning.
+    /// It does this via running `read_bytes` with each chunk with `query_timeout` set.
+    #[instrument(skip_all, level = "trace")]
+    pub async fn upload_and_verify(&self, bytes: Bytes) -> Result<ChunkAddress> {
+        self.upload_bytes(bytes, true).await
+    }
+
+    // --------------------------------------------
+    // ---------- Private helpers -----------------
+    // --------------------------------------------
+
+    #[instrument(skip(self, bytes), level = "trace")]
+    async fn upload_bytes(&self, bytes: Bytes, verify: bool) -> Result<ChunkAddress> {
+        match LargeFile::new(bytes.clone()) {
+            Ok(file) => self.upload_large(file, verify).await,
+            Err(Error::TooSmallForSelfEncryption { .. }) => {
+                let file = SmallFile::new(bytes)?;
+                self.upload_small(file, verify).await
+            }
+            Err(error) => Err(error)?,
+        }
+    }
+
+    /// Directly writes a [`SmallFile`] to the network in the
+    /// form of a single chunk, without any batching.
+    #[instrument(skip_all, level = "trace")]
+    async fn upload_small(&self, small: SmallFile, verify: bool) -> Result<ChunkAddress> {
+        let chunk = package_small(small)?;
+        let address = *chunk.address();
+
+        if self.verify_chunk_is_stored(address).await.is_ok() {
+            return Ok(address);
+        }
+
+        self.store_chunk(chunk).await?;
+
+        if verify {
+            self.verify_chunk_is_stored(address).await?;
+        }
+
+        Ok(address)
+    }
+
+    /// Directly writes a [`LargeFile`] to the network in the
+    /// form of immutable self encrypted chunks, without any batching.
+    #[instrument(skip_all, level = "trace")]
+    async fn upload_large(&self, large: LargeFile, verify: bool) -> Result<ChunkAddress> {
+        let (head_address, all_chunks) = encrypt_large(large)?;
+        for next_batch in all_chunks.chunks(CHUNKS_BATCH_MAX_SIZE) {
```

Comment:
Here again, if we got rid of the Small/Large distinction we would simplify the code.

Reply:
Yep.